### ISSUE TITLE
Provide a convenient way for the user to reset the grad

### ISSUE THREAD

**Issue by vtantia** [pytorch/pytorch#41696]
## 🚀 Feature
Provide a convenient way for the user to reset the grad

## Motivation
Resetting the grad instead of zero-ing it can save us extra memory between `optimizer.zero_grad()` and the next `loss.backward()`

## Pitch
A new function in `torch/optim/optimizer.py`, maybe called `reset_grad()`, which sets `p.grad` to `None`

## Alternatives
An alternative we considered is to allow the user to define their own helper function to do this, however, it seems like a feature which could be used by advanced users

## Additional context
We are implementing a distributed optimization algorithm [SlowMo](https://arxiv.org/pdf/1910.00643.pdf), which requires extra memory between `optimizer.zero_grad()` and the next `loss.backward()` call, and thought this might be a good way to save memory for it.

with @albanD 

cc @ezyang @SsnL @albanD @zou3519 @gqchen @vincentqb

**Comment by ngimel** [comment]
This not only saves memory, but makes gradient accumulation more efficient (if `.grad` is None, the computed gradient can be directly moved into `.grad` attribute, without any memcopies or unnecessary addition operations). 
I support having this function in the core, for now the workaround is to do it yourself with a helper function.

**Comment by firstprayer** [comment]
Having both zero_grad() and reset_grad() might be kind of confusing to the end users. 

Are zero-ing all gradients always equivalent to setting all gradients to None? Wonder if there's any case where zero_grad() is preferred over reset_grad().

If the two are not equivalent, and have pros and cons, we probably should mention those pros and cons in the documentations to articulate the difference in order to prevent confusion. We should probably have some default recommendation (e.g. do we think reset_grad() will be better in most cases, which looks like the case, because most references I found are just calling .backward() right after calling zero_grad()).

**Comment by albanD** [comment]
> Are zero-ing all gradients always equivalent to setting all gradients to None?

Unfortunately, it is not.
The main cases where it appears is:
- `nn.optim` optimizers have a different behavior if the gradient is 0 or None (in one case it does the step with a gradient of 0 and in the other it skip the step altogether).
- When user try to access the gradient value and perform manual ops on it. A None attribute or a Tensor full of 0s will be different.
- The few cases where we check if `.grad` is None as a hint if the backward pass touched this Tensor or not (in autograd.grad or Tensor.grad warning for example). Note that, in this case, this won't make it more wrong, but it will be BC-breaking.

Unfortunately, the pros and cons here are quite tricky and depend on a lot on the model/device you run on. So it might be tricky to provide a good general advice to the users.
Also changing the default of `.zero_grad()` will be very hard because of BC-breaking implications.

But having the two functions and doc that points from one to the other with the detail of what they do should work?

**Comment by mcarilli** [comment]
I definitely think setting grads to None should be a first-class citizen, for all the reasons mentioned above.  I've already described a manual approach here https://pytorch.org/docs/stable/autograd.html#default-gradient-layouts, which is valid, but more verbose than it needs to be.

From an API perspective, an optional kwarg to `zero_grad` instead of `reset_grad` might be easier for users to plumb through.  Users could write
```python
opt.zero_grad(set_to_none=args.set_grads_none)
```
instead of
```python
if args.set_grads_none:
    opt.reset_grad()
else:
    opt.zero_grad()
```
But having a separate `reset_grad call` emphasizes that it's a separate concept.  Really a matter of style.  As long as setting grads to None is documented as a first class citizen somehow, im happy.

**Comment by vincentqb** [comment]
> This not only saves memory, but makes gradient accumulation more efficient (if `.grad` is None, the computed gradient can be directly moved into `.grad` attribute, without any memcopies or unnecessary addition operations).

This is a very good reason to support it.

> * `nn.optim` optimizers have a different behavior if the gradient is 0 or None (in one case it does the step with a gradient of 0 and in the other it skip the step altogether).

Once the user invokes `loss.backward()`, all gradients should be populated with non-None values and optimizers would then behave the same, no?

> * When user try to access the gradient value and perform manual ops on it. A None attribute or a Tensor full of 0s will be different.

Yes, to me this is a strong reason not to change the default behavior of `zero_grad`.

> * The few cases where we check if `.grad` is None as a hint if the backward pass touched this Tensor or not (in autograd.grad or Tensor.grad warning for example). Note that, in this case, this won't make it more wrong, but it will be BC-breaking.

Can you clarify "more wrong" ? :)

> Having both zero_grad() and reset_grad() might be kind of confusing to the end users.

I agree with this, and I can even imagine some users incorrectly believing they need both. Passing a parameter to `zero_grad` as mentioned above looks reasonable to me:
```
opt.zero_grad(set_to_none=True)
```
Thoughts?

**Comment by albanD** [comment]
> Once the user invokes loss.backward(), all gradients should be populated with non-None values and optimizers would then behave the same, no?

Yes as long as all the weights were used to compute the loss.
If some were not, then their `.grad` will stay as `None`. While it would remain at `0` before.

> Can you clarify "more wrong" ? :)

The short answer is that in the autograd today, things that are independent or have zero gradient are not differentiated.
And these two cases can manifest themselves as either a Tensor full of 0s, None or an error (this last one when using `autograd.grad(..., allow_unused=False)`).
I mean it won't be "more wrong" in the sense that it still follows this rule. But one could argue that it is not a great rule haha

**Comment by ngimel** [comment]
Fixed by #44423
